File: gdb/memory.py
#/usr/bin/python
#
##################################################################################
#
# MIT License
#
# Copyright (c) 2022 AO Kaspersky Lab. All Rights Reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#
##################################################################################
#
# Memory utilities
#

import chunk
import struct
import sys
from functools import lru_cache
import gdb

def CastULong(v):
    mask = (1 << 64) - 1
    return int(v.cast(gdb.Value(mask).type)) & mask

def AlignAddress(address, align):
    return address + ((align - (address % align)) % align)

@lru_cache()
def ReadMemory_Int(addr):
    """Return an integer read from memory."""
    sz = 8
    mem = gdb.selected_inferior().read_memory(addr, sz).tobytes()
    return struct.unpack("{}Q".format('<'), mem)[0]

@lru_cache()
def LookupType(name):
    try:
        return gdb.lookup_type(name).strip_typedefs()
    except RuntimeError:
        return None

def FindHeapBase():
    return int(gdb.parse_and_eval("mp_->sbrk_base"))

def FindMainArena():
    return AlignAddress(CastULong(gdb.parse_and_eval("(void *)&__malloc_hook")) + 8, 0x20)

class Arena:
    """Heap arena class"""

    def __init__(self, addr, malloc_state_type):
        arena = gdb.parse_and_eval("*{:#x}".format(addr))
        self.m_addr = int(arena.address)
        self.m_arena = arena.cast(malloc_state_type)
        return

    def __getitem__(self, item):
        return self.m_arena[item]

    def __getattr__(self, attr):
        return self.m_arena[attr]

def GetMainArena():
    malloc_state_type = LookupType("struct malloc_state")
    main_arena = FindMainArena()
    return Arena(main_arena, malloc_state_type)


class Chunk:
    """Heap chunk class"""

    def __init__(self, addr):
        self.ptrsize = 8
        self.base_address = addr
        self.address = addr + 2 * self.ptrsize
        self.size_addr = int(self.address - self.ptrsize)
        self.prev_size_addr = self.base_address
        return

    @property
    def size(self):
        return ReadMemory_Int(self.size_addr) & (~0x07)

    @property
    def user_size(self):
        current_size = self.size
        if current_size == 0:
            return current_size
        if self.m_bit:
            return current_size - 2 * self.ptrsize
        return current_size - self.ptrsize

    @property
    def prev_chunk_size(self):
        return ReadMemory_Int(self.prev_size_addr)

    def get_next_chunk(self):
        return Chunk(self.base_address + self.size)

    @property
    def p_bit(self):
        return ReadMemory_Int(self.size_addr) & 0x01 == 0x01

    @property
    def m_bit(self):
        return ReadMemory_Int(self.size_addr) & 0x02 == 0x02

    @property
    def a_bit(self):
        return ReadMemory_Int(self.size_addr) & 0x04 == 0x04

    @property
    def busy(self):
        if self.m_bit:
            return True
        return self.get_next_chunk().p_bit

def GetChunks():
    addr = FindHeapBase()
    main_arena = GetMainArena()
    while True:
        chunk = Chunk(addr)
        yield chunk
        if chunk.size == 0:
            break
        addr += chunk.size
        if addr >= main_arena.top:
            break


def AddItem(d, val):
    if d.get(val) is None:
        d[val] = 1
    else:
        d[val] = d[val] + 1

def SplitStringIntoTwixes(s):
    for pos in range(len(s), 1, -2):
        i = int(s[pos-2:pos], 16)
        yield chr(i) if i > 21 else '.'

def HexIntToAscii(v):
    return "".join(SplitStringIntoTwixes(v))

def PrintTopStats(d):
    s = sorted(d.items(), key=lambda x: x[1], reverse=True)
    i = 0
    for k, v in s:
        print("{} - '{}': {}".format(k, HexIntToAscii(k), v))
        i = i + 1
        if i > 10:
            break

class HeapAllocStats(gdb.Command):
    """Collect heap allocation stats"""

    def __init__ (self):
        super (HeapAllocStats, self).__init__ ("heap_alloc_stats", gdb.COMMAND_USER)

    def invoke (self, arg, from_tty):
        if gdb.selected_inferior().pid == 0:
            print("There is no running program\n")
            return

        stat_int1 = dict();
        stat_int2 = dict();
        stat_int3 = dict();
        stat_int4 = dict();
        stat_double_int1 = dict();
        stat_double_int2 = dict();
        stat_quad_int = {};
        busy_chunk_count = 0

        for ch in GetChunks():
            if ch.busy:
                busy_chunk_count = busy_chunk_count + 1
                size = ch.user_size
                i1 = 0
                i2 = 0
                i3 = 0
                i4 = 0
                if size >= 32:
                    i1 = ReadMemory_Int(ch.address)
                    i2 = ReadMemory_Int(ch.address + 8)
                    i3 = ReadMemory_Int(ch.address + 16)
                    i4 = ReadMemory_Int(ch.address + 24)
                elif size >= 24:
                    i1 = ReadMemory_Int(ch.address)
                    i2 = ReadMemory_Int(ch.address + 8)
                    i3 = ReadMemory_Int(ch.address + 16)
                elif size >= 16:
                    i1 = ReadMemory_Int(ch.address)
                    i2 = ReadMemory_Int(ch.address + 8)
                elif size >= 8:
                    i1 = ReadMemory_Int(ch.address)
                else:
                    continue
                AddItem(stat_int1, "{:016x}".format(i1))
                AddItem(stat_int2, "{:016x}".format(i2))
                AddItem(stat_int3, "{:016x}".format(i3))
                AddItem(stat_int4, "{:016x}".format(i4))
                AddItem(stat_double_int1, "{:016x}{:016x}".format(i2, i1))
                AddItem(stat_double_int2, "{:016x}{:016x}".format(i4, i3))
                AddItem(stat_quad_int, "{:016x}{:016x}{:016x}{:016x}".format(i4, i3, i2, i1))
        print("Total {} busy chunks\n".format(busy_chunk_count))

        print("--- Top by first int ---");
        PrintTopStats(stat_int1);
        print("\n--- Top by second int ---\n");
        PrintTopStats(stat_int2);
        print("\n--- Top by third int ---\n");
        PrintTopStats(stat_int3);
        print("\n--- Top by fourth int ---\n");
        PrintTopStats(stat_int4);
        print("\n--- Top by first double int ---\n");
        PrintTopStats(stat_double_int1);
        print("\n--- Top by second double int ---\n");
        PrintTopStats(stat_double_int2);
        print("\n--- Top by quad int ---\n");
        PrintTopStats(stat_quad_int);


if __name__ == "__main__":

    if sys.version_info[0] == 2:
        err("Python3 is required")

    else:
        gdb.execute("set python print-stack full")

        HeapAllocStats()

File: memory.js
/*///////////////////////////////////////////////////////////////////////
MIT License

Copyright (c) 2021-2022 AO Kaspersky Lab. All Rights Reserved.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
/*////////////////////////////////////////////////////////////////////////

/////////////////////////////////////////////////////////////////////////
// Memory utilities
//
// Note! Parameters should be passed with 0x or in "". Otherwise you got an error "Error: Unable to bind name ...".
//
// !av_heap_alloc_stats <heap address>, <block size in hex> - walk list of memory blocks of specified size in heap and collect stats about allocation stacks.
//
// !heap_alloc_stats <heap_address> - walk list of memory blocks in the heap and collect stats about the contents
//
/////////////////////////////////////////////////////////////////////////

"use strict";

// Utilities
function GetStringParameter(param)
{
    if (typeof param == "object")
    {
        return param.toString(16);
    }

    return "" + param.toString(16);
}

function PrintCommandOutput(output)
{
    for (var line of output)
        host.diagnostics.debugLog(line + "\n");
}

function* CreateCacheReader(filename)
{
    var reader = host.namespace.Debugger.Utility.FileSystem.CreateTextReader(filename, "Utf8");
    try
    {
        for (var line = reader.ReadLine(); line.length > 0; line = reader.ReadLine())
        {
            yield line;
        }
    }
    catch(err)
    {
        if (err.message != "Cannot read past end of file") {
            host.diagnostics.debugLog("Cache read error: " + err.message + "\n");
        }
        return "";
    }

    return "";
}

function GetDataForProcessing(heap)
{
    var heapAddress = GetStringParameter(heap);
    var heapCacheFilename = "heap_" + heapAddress + ".cache";

    // cache filename have to depends on dump filename
    try {
        if (host.namespace.Debugger.Sessions[0].Attributes.Target.Details.DumpFileName.length > 0) {
            heapCacheFilename = host.namespace.Debugger.Sessions[0].Attributes.Target.Details.DumpFileName + "_" + heapCacheFilename;
        }
    } catch (err) {
        host.diagnostics.debugLog("Dump name getting error: " + err.message + "\n");
    }

    if (host.namespace.Debugger.Utility.FileSystem.FileExists(heapCacheFilename)) {
        return CreateCacheReader(heapCacheFilename);
    }

    var output = host.namespace.Debugger.Utility.Control.ExecuteCommand("!heap -s -a -c -h " + heapAddress);
    var cacheFile = host.namespace.Debugger.Utility.FileSystem.CreateFile(heapCacheFilename, "CreateAlways");
    var writer = host.namespace.Debugger.Utility.FileSystem.CreateTextWriter(cacheFile, "Utf8");

    var errorCount = 0;
    var writeLineCount = 0;
    for (let line of output) {
        ++writeLineCount;
        try {
            if (line.length > 0) {
                writer.WriteLine(line);
            }
        } catch (err) {
            if (errorCount < 20) {
                host.diagnostics.debugLog("Cache write error at line " + writeLineCount.toString(10) + ": " + err.toString() + "\n");
            }
            ++errorCount;
        }
    }
    cacheFile.Close();
    host.diagnostics.debugLog("Total write errors: " + errorCount.toString(10) + "\n");
    return output;
}

function AddItem(stat, key)
{
    if (stat.has(key)) {
        stat.set(key, stat.get(key) + 1);
    } else {
        stat.set(key, 1);
    }
}

function __CollectAvHeapAllocationStats(heap, blockSize)
{
    var stat = new Map();
    var lineCount = 0;
    var output = GetDataForProcessing(heap);

    var regExpText = "^([0-9A-Fa-f]{8})\\sB\\sabcdaaa[0-9A-Fa-f]\\s[0-9A-Fa-f]{8}\\s" + ('0000000' + blockSize.toString(16)).slice(-8) + "\\s(.+)$";
    //host.diagnostics.debugLog("RegExp: " + regExpText + "\n");
    var lineRE = new RegExp(regExpText);

    var matched = 0;
    try
    {
        for (var line of output)
        {
            //host.diagnostics.debugLog("block " + line + "\n");
            ++lineCount;

            var parsed = line.match(lineRE);
            if (parsed) {
                ++matched;

                var address = host.parseInt64(parsed[1], 16);
                var stackPtr = host.parseInt64(host.memory.readMemoryValues(address + 0x20, 1, 4), 10);
                //host.diagnostics.debugLog("  " + address.toString(16) + " - " + stackPtr.toString(16) + "\n");
                AddItem(stat, stackPtr);
            }

        }
    }
    catch(err)
    {
        //host.diagnostics.debugLog("Error: " + err.toString() + "\n");
    }

    host.diagnostics.debugLog("Matched: 0x" + matched.toString(16) + " blocks\n");
    host.diagnostics.debugLog("-------------------------------------------------------------\n");

    // iterate map sorted by value
    stat[Symbol.iterator] = function* () {
        yield* [...this.entries()].sort((a, b) => b[1] - a[1]);
    }

    for (let [key, value] of stat) {
        host.diagnostics.debugLog("\n=== stack at 0x" + ('0000000' + key.toString(16)).slice(-8) + ' - ' + value.toString(10) + " blocks ===\n");

        var output = host.namespace.Debugger.Utility.Control.ExecuteCommand("dds " + (key + 0xC).toString(16) + " L20");
        PrintCommandOutput(output);
    }

    host.diagnostics.debugLog("\n");
    return lineCount;
}

function hex2ascii(str)
{
    var hex = str.toString();
    var res = '';
    for (var i = hex.length - 2; i >= 0; i -= 2) {
        var code = parseInt(hex.substr(i, 2), 16);
        if (code < 21) {
            res += ".";
        } else {
            res += String.fromCharCode(code);
        }
    }
    return res;
}

function PrintTopStats(stat)
{
    // iterate map sorted by value
    stat[Symbol.iterator] = function* () {
        yield* [...this.entries()].sort((a, b) => b[1] - a[1]);
    }

    var i = 0;
    for (let [key, value] of stat) {
        host.diagnostics.debugLog(key + " \"" + hex2ascii(key) + "\": " + value.toString(10) + "\n");
        if (++i > 10) {
            break;
        }
    }
}

function __CollectHeapAllocationStats(heap) {
    var stat_Dword1 = new Map();
    var stat_Dword2 = new Map();
    var stat_Dword3 = new Map();
    var stat_Dword4 = new Map();
    var stat_Qword1 = new Map();
    var stat_Qword2 = new Map();
    var lineCount = 0;
    var output = GetDataForProcessing(heap);

    var regExpText = "^[0-9A-Fa-f]{8,16}\\s[bB]\\s([0-9A-Fa-f]{8})\\s([0-9A-Fa-f]{8})\\s([0-9A-Fa-f]{8})\\s([0-9A-Fa-f]{8})\\s(.+)$";
    //host.diagnostics.debugLog("RegExp: " + regExpText + "\n");
    var lineRE = new RegExp(regExpText);

    var matched = 0;
    try {
        for (var line of output) {
            //host.diagnostics.debugLog("block " + line + "\n");
            ++lineCount;

            var parsed = line.match(lineRE);
            if (parsed) {
                ++matched;
                //host.diagnostics.debugLog("Parsed\n  " + parsed[1] + "\n  " + parsed[2] + "\n  " + parsed[3] + "\n  " + parsed[4] + "\n");

                AddItem(stat_Dword1, parsed[1]);
                AddItem(stat_Dword2, parsed[2]);
                AddItem(stat_Dword3, parsed[3]);
                AddItem(stat_Dword4, parsed[4]);
                AddItem(stat_Qword1, parsed[2] + parsed[1]);
                AddItem(stat_Qword2, parsed[4] + parsed[3]);
            }
        }
    }
    catch (err) {
        //host.diagnostics.debugLog("Error: " + err.toString() + "\n");
    }

    host.diagnostics.debugLog("Matched: 0x" + matched.toString(16) + " blocks\n");
    
    host.diagnostics.debugLog("\n--- Top by first dword ---\n");
    PrintTopStats(stat_Dword1);
    host.diagnostics.debugLog("\n--- Top by second dword ---\n");
    PrintTopStats(stat_Dword2);
    host.diagnostics.debugLog("\n--- Top by third dword ---\n");
    PrintTopStats(stat_Dword3);
    host.diagnostics.debugLog("\n--- Top by fourth dword ---\n");
    PrintTopStats(stat_Dword4);
    host.diagnostics.debugLog("\n--- Top by first qword ---\n");
    PrintTopStats(stat_Qword1);
    host.diagnostics.debugLog("\n--- Top by second qword ---\n");
    PrintTopStats(stat_Qword2);

    host.diagnostics.debugLog("\n");
    return lineCount;
}

// Register aliases
function initializeScript()
{
    return [new host.apiVersionSupport(1, 2),
        new host.functionAlias(__CollectAvHeapAllocationStats, "av_heap_alloc_stats"),
        new host.functionAlias(__CollectHeapAllocationStats, "heap_alloc_stats")];
}
